File: web/backend/app/api/backtest.py
# ============================================================
# NEXUS TRADER Web — Backtest Router
#
# POST /backtest/start            — launch a backtest job
# GET  /backtest/status/{job_id}  — poll progress
# GET  /backtest/results/{job_id} — completed results
# POST /backtest/cancel/{job_id}  — cancel a running job (Phase 5)
# ============================================================
from __future__ import annotations

import logging
from typing import List, Optional

from fastapi import APIRouter, Depends, Path, Query
from pydantic import BaseModel

from app.auth.dependencies import get_current_user
from app.api.engine import _send_engine_command

logger = logging.getLogger(__name__)
router = APIRouter(
    prefix="/backtest",
    tags=["backtest"],
    dependencies=[Depends(get_current_user)],
)


class BacktestRequest(BaseModel):
    symbols: List[str] = ["BTC/USDT", "ETH/USDT", "SOL/USDT"]
    start_date: str = "2024-01-01"
    end_date: str = "2026-03-01"
    timeframe: str = "30m"
    fee_pct: float = 0.04


@router.post("/start")
async def start_backtest(req: BacktestRequest):
    """Launch a backtest job asynchronously."""
    return await _send_engine_command("start_backtest", {
        "symbols": req.symbols,
        "start_date": req.start_date,
        "end_date": req.end_date,
        "timeframe": req.timeframe,
        "fee_pct": req.fee_pct,
    }, timeout=15)


@router.get("/status/{job_id}")
async def backtest_status(job_id: str = Path(description="Backtest job ID")):
    """Poll backtest progress."""
    return await _send_engine_command("get_backtest_status", {"job_id": job_id})


@router.get("/results/{job_id}")
async def backtest_results(job_id: str = Path(description="Backtest job ID")):
    """Get completed backtest results."""
    return await _send_engine_command("get_backtest_results", {"job_id": job_id})


@router.post("/cancel/{job_id}")
async def cancel_backtest(job_id: str = Path(description="Backtest job ID to cancel")):
    """Cancel a running backtest job."""
    return await _send_engine_command("cancel_backtest", {"job_id": job_id})

File: web/frontend/src/api/backtest.ts
import api from './client';

export interface BacktestParams {
  symbols: string[];
  start_date: string;
  end_date: string;
  timeframe: string;
  fee_pct: number;
}

export interface BacktestStartResponse {
  job_id: string;
  status: string;
}

export interface BacktestStatus {
  job_id: string;
  status: string;
  progress_pct: number;
  elapsed_s: number;
}

export interface BacktestMetrics {
  pf: number;
  wr: number;
  max_dd: number;
  cagr: number;
  n_trades: number;
  sharpe: number;
}

export interface BacktestResults {
  job_id: string;
  status: string;
  metrics: BacktestMetrics;
  trades: Array<Record<string, unknown>>;
}

export async function startBacktest(params: BacktestParams): Promise<BacktestStartResponse> {
  const resp = await api.post('/backtest/start', params);
  const d = resp.data;
  return { job_id: d.job_id, status: d.status || 'ok' };
}

export async function getBacktestStatus(jobId: string): Promise<BacktestStatus> {
  const resp = await api.get(`/backtest/status/${jobId}`);
  const d = resp.data;
  return {
    job_id: jobId,
    status: d.state || d.status || 'unknown',
    progress_pct: d.progress_pct ?? d.progress ?? 0,
    elapsed_s: d.elapsed_s ?? 0,
  };
}

export async function getBacktestResults(jobId: string): Promise<BacktestResults> {
  const resp = await api.get(`/backtest/results/${jobId}`);
  const d = resp.data;
  const r = d.results || {};
  return {
    job_id: jobId,
    status: d.state || d.status || 'unknown',
    metrics: {
      pf: r.profit_factor ?? 0,
      wr: r.win_rate ?? 0,
      max_dd: r.max_drawdown ?? 0,
      cagr: r.cagr ?? 0,
      n_trades: r.total_trades ?? 0,
      sharpe: r.sharpe ?? 0,
    },
    trades: r.trades || [],
  };
}
